File: Data Structures & Algorithms/subarray-sum-equals-k/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number}
     */
    subarraySum(nums, k) {
        let res = 0
        let currsum = 0
        const prefix_map = new Map()
        prefix_map.set(0, 1)
        for ( const num of nums) {
            currsum += num
            const diff = currsum - k

            res += prefix_map.get(diff) || 0;
            prefix_map.set(currsum, 1 + (prefix_map.get(currsum) || 0))
        }

        return res
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1 for num in nums]
        suffix = [1 for num in nums]

        for i in range(1, len(prefix)):
            prefix[i] = prefix[i - 1] * nums[i - 1]

        for i in range(len(suffix) - 2, -1, -1):
            suffix[i] = suffix[i + 1] * nums[i + 1]

        return [suffix[i] * prefix[i] for i in range(len(nums))]